Initial work on adding Dynamic Type support

spantaleev opened the pull request:
Related to issue #496.

This patch is not exhaustive - it doesn't make everything use Dynamic Type fonts.
Still, it makes most of the (frequently used) UI scalable, which should be a nice first step and an improvement for people with visibility problems.

Even in the generic case (no visibility problems and font size increase set in Accessibility settings), with this patch, the fonts appear a little larger and more legible than before (addresses issue #1487).

This patch defines some basic fonts, listens for a `UIContentSizeCategoryDidChangeNotification` and maintains these fonts updated accordingly.

To let the rest of the app know about such font size changes, we send the same `kRiotDesignValuesDidChangeThemeNotification` which gets dispatched when people change between the Dark/Light themes. Since the intention is similar (UI settings update & redraw), it seemed natural to repurpose it.

Not all fonts are assigned dynamically (`widget.font = ..`), so not all widgets use the new font values from `RiotDesignValues`.
In other places, we simply configure the Dynamic Type font inside the `.xib` file.

UI widgets which have their font defined in the `.xib` file do not get immediately updated. Using the "Automatically Adjust Font" property seems like an easy viable solution, but it requires iOS>=10, so it hasn't been done.

Signed-off-by: Slavi Pantaleev <s.pantaleev@gmail.com>

ara4n commented:
@spantaleev thanks for this! @giomfo or @manuroe please can you take a look? I know a certain ex-boss who would love for this to land...

giomfo commented on Riot/Views/RoomInputToolbar/RoomInputToolbarView.m:
```diff
@@ -78,18 +78,22 @@ - (void)awakeFromNib
-    
-    growingTextView.font = [UIFont systemFontOfSize:15];
+
+    growingTextView.font = kRiotTextFontBody;
```
About `RoomInputToolbarView`, I agree to apply this change, but the corresponding xib file must be adjusted. Indeed the icons are not aligned when the user selects a large text size.
See the following screenshot with the view frames:

![screen shot 2017-10-30 at 19 15 09](https://user-images.githubusercontent.com/8969772/32187990-d438104a-bda6-11e7-9603-2407344b1b00.png)

I know this is not your fault ;). We may update `RoomInputToolbarView` class in another PR if you prefer.

giomfo commented on Riot/Views/RoomBubbleList/RoomIncomingTextMsgBubbleCell.xib:
```diff
@@ -1,11 +1,11 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="11762" systemVersion="15G1108" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" colorMatched="YES">
+<?xml version="1.0" encoding="UTF-8"?>
```
The room bubble views are very tricky, we should not update this kind of view in this first PR. We will handle them later in another PR.

Please remove any changes applied inside the "RoomBubbleList" directory for this PR.

As you can see in the following screenshot with a large text size, you forgot to update some views (small username), and updated views layout is not correct (username and message are too close):

![screen shot 2017-10-30 at 19 22 10](https://user-images.githubusercontent.com/8969772/32188299-b4a6b546-bda7-11e7-8c4d-79d88be753f9.png)

giomfo commented on Riot/Views/RoomBubbleList/RoomOutgoingTextMsgBubbleCell.xib:
```diff
@@ -1,11 +1,11 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="11762" systemVersion="15G1108" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" colorMatched="YES">
+<?xml version="1.0" encoding="UTF-8"?>
```
see my comment above

giomfo commented on Riot/Views/RoomList/PublicRoomTableViewCell.xib:
```diff
@@ -1,11 +1,11 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="11762" systemVersion="15G1108" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" colorMatched="YES">
+<?xml version="1.0" encoding="UTF-8"?>
```
Note: During my tests, I observed the dynamic text size was very well supported in this cell.
We should use this xib file as reference to adjust the other cells in the coming PRs.

giomfo commented on Riot/Views/RoomList/RoomCollectionViewCell.m:
```diff
@@ -123,16 +123,8 @@ - (void)render:(MXKCellData *)cellData
                 
```
It is too early to update the `RoomCollectionViewCell` class.
Please remove all the changes for this class in this PR.

The display with a large text size is not relevant:
![screen shot 2017-10-30 at 19 40 15](https://user-images.githubusercontent.com/8969772/32189151-32439c6a-bdaa-11e7-9cbd-9a475fc3e4e3.png)

giomfo commented on Riot/Views/RoomList/RoomTableViewCell.xib:
```diff
@@ -17,7 +17,7 @@
             <autoresizingMask key="autoresizingMask"/>
             <tableViewCellContentView key="contentView" opaque="NO" clipsSubviews="YES" multipleTouchEnabled="YES" contentMode="center" tableViewCell="L2L-l5-wPx" id="aXz-IR-jj5">
-                <rect key="frame" x="0.0" y="0.0" width="600" height="74"/>
+                <rect key="frame" x="0.0" y="0.0" width="600" height="73.5"/>
```
I think we should keep 74 for the height. 
No change seems required for this xib file

giomfo commented on Riot/Views/RoomTitle/ExpandedRoomTitleView.xib:
```diff
@@ -38,31 +38,31 @@
                     </constraints>
```
I would prefer update `ExpandedRoomTitleView` class later in another PR.
Please remove your changes for the moment.

giomfo commented on Riot/Views/RoomTitle/PreviewRoomTitleView.xib:
```diff
@@ -6,7 +6,6 @@
     <dependencies>
```
I would prefer update `PreviewRoomTitleView` class later in another PR.
Please remove your changes for the moment.

giomfo commented on Riot/Views/RoomTitle/RoomTitleView.xib:
```diff
@@ -19,7 +19,7 @@
                     <color key="backgroundColor" red="0.0" green="0.0" blue="0.0" alpha="0.0" colorSpace="custom" customColorSpace="sRGB"/>
                     <accessibility key="accessibilityConfiguration" identifier="DisplayNameTextField"/>
-                    <fontDescription key="fontDescription" type="system" weight="medium" pointSize="18"/>
+                    <fontDescription key="fontDescription" style="UICTFontTextStyleHeadline"/>
```
I would prefer update `RoomTitleView` class later in another PR.
Please remove your changes for the moment.

spantaleev commented:
@giomfo, thanks for the review!

I've made the changes as per your advice and rebased my changes on top of `develop`.

Some notes about what wasn't done:

- about the bing font, I'm not sure how we can have a medium-weight in a Dynamic Type system. Maybe we should do this in another PR later, if there turns out to be some way of doing it.

- I'm not sure how to properly align the boxes in `RoomInputToolbarView`. Maybe we should do it later in another PR.

giomfo commented:
Hi @spantaleev, I don't see your changes, did you push them?

spantaleev commented:
@giomfo, it should be `--force` pushed to the same branch.

As the history in #496 indicates:
- my first push was for [dd8ab16a04b7391b](https://github.com/spantaleev/riot-ios/commit/dd8ab16a04b7391b44a9711fb80a121515e6476f), which you have reviewed above
- since then, I've made: changes based on your feedback, rebased against `develop` and `--force` pushed to the same branch (`add-dynamic-type`), which became commit [05027d64c0acd64](https://github.com/spantaleev/riot-ios/commit/05027d64c0acd64570c130b25ff5afc801693377)

giomfo commented on Riot/Views/RoomList/RoomCollectionViewCell.xib:
```diff
@@ -1,11 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Please revert your changes in this .xib (like you did in RoomCollectionViewCell.m file).
We will work on this collection view cell later

spantaleev commented:
@giomfo, I've just done the following:
- removed the changes done to `RoomCollectionViewCell.xib`
- rebased on top of `develop` again

I guess you could easily take this `dynamic-type` branch now and continue forward.

giomfo commented:
Thank @spantaleev to have initiated the Dynamic Type support.
Feel free to provide new PRs on Dynamic Type support by considering the`vector-im:dynamic-type` branch

gerroon commented:
Hi

Is this in? If so how can I use it? The fonts are so small

QEDeD commented:
@spantaleev would you by any chance be able to take a look at implementing Dynamic Type again? Sadly it hasn't been properly implemented yet.

spantaleev commented:
No.. Doing it once was enough. I'm not even an iOS developer, so it took some effort and annoyance dealing with the tools.. In the end, it didn't even get merged, so there's no point in doing it again.

QEDeD commented:
@spantaleev completely understandable, thanks for your contribution regardless :)

johnwaldo commented:
@giomfo it looks like the dynamic-type branch has been deleted from the element ios tree. Can you provide an update on the effort to add support for dynamic type on ios? This seems like a huge accessibility/readability issue that needs attention. 

@spantaleev thank you for your efforts!!
